client/tso_service_discovery.go

```diff
@@ -526,3 +503,83 @@ func (c *tsoServiceDiscovery) findGroupByKeyspaceID(
 
 	return resp.KeyspaceGroup, nil
 }
+
+func (c *tsoServiceDiscovery) getTSOServer(sd ServiceDiscovery) (string, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	var (
+		addrs []string
+		err   error
+	)
+	t := c.tsoServerDiscovery
+	if len(t.addrs) == 0 || t.failureCount == len(t.addrs) {
+		addrs, err = sd.DiscoverMicroservice(tsoService)
+		if err != nil {
+			return "", err
+		}
+		failpoint.Inject("serverReturnsNoTSOAddrs", func() {
+			addrs = nil
+		})
+		if len(addrs) == 0 {
+			// There is no error but no tso server address found, which means
+			// either the server side is experiencing some problems to get the
+			// tso primary addresses or the server hasn't been upgraded to the
+			// version which processes and returns GetClusterInfoResponse.TsoUrls.
+			// In this case, we fall back to the old way of discovering the tso
+			// primary addresses from etcd directly.
+			log.Warn("[tso] no tso server address found,"+
+				" fallback to the legacy path to discover from etcd directly",
+				zap.String("discovery-key", c.defaultDiscoveryKey))
+			addrs, err = c.discoverWithLegacyPath()
```

Comment:
if api server has been ready without tso service, but tso server is starting, what will happen? return error "not found"?

Reply:
It returns communication error or "not started" or "not served" for this case.

Reply:
"no error and tso_urls is nil" indicates the server version is behind and doesn't return tso_urls.